ext/liquid_c/variable.c

```diff
@@ -50,37 +35,67 @@ static VALUE try_variable_strict_parse(VALUE uncast_args)
+        VALUE const_keyword_args = Qnil;
+        VALUE push_keywords_obj = Qnil;
+        vm_assembler_t *push_keywords_code = NULL;
 
         if (parser_consume(&p, TOKEN_COLON).type) {
             do {
                 if (p.cur.type == TOKEN_IDENTIFIER && p.next.type == TOKEN_COLON) {
                     VALUE key = token_to_rstr(parser_consume_any(&p));
                     parser_consume_any(&p);
 
-                    if (const_keyword_args && !will_parse_constant_expression_next(&p))
-                        const_keyword_args = false;
-                    if (keyword_args == Qnil)
-                        keyword_args = rb_hash_new();
-                    rb_hash_aset(keyword_args, key, parse_expression(&p));
+                    keyword_arg_count++;
+
+                    if (push_keywords_obj != Qnil) {
+                        vm_assembler_add_push_const(push_keywords_code, key);
+                        parse_and_compile_expression(&p, push_keywords_code);
+                    } else {
+                        VALUE const_obj = try_parse_constant_expression(&p);
+
+                        if (const_obj != Qundef) {
+                            if (const_keyword_args == Qnil)
+                                const_keyword_args = rb_hash_new();
+                            rb_hash_aset(const_keyword_args, key, const_obj);
+                        } else {
+                            expression_t *push_keywords_expr;
+                            // use an object to automatically free on an exception
+                            push_keywords_obj = expression_new(&push_keywords_expr);
+                            rb_obj_hide(push_keywords_obj);
+                            push_keywords_code = &push_keywords_expr->code;
+
+                            if (const_keyword_args != Qnil) {
+                                rb_hash_foreach(const_keyword_args, push_each_const_keyword_arg, (VALUE)push_keywords_code);
+                                const_keyword_args = Qnil;
+                            }
+                            vm_assembler_add_push_const(push_keywords_code, key);
+                            parse_and_compile_expression(&p, push_keywords_code);
+                        }
+                    }
                 } else {
                     parse_and_compile_expression(&p, code);
                     arg_count++;
                 }
             } while (parser_consume(&p, TOKEN_COMMA).type);
         }
 
-        if (keyword_args != Qnil) {
+        if (keyword_arg_count) {
             arg_count++;
-            if (RHASH_SIZE(keyword_args) > 255) {
+            if (keyword_arg_count > 255)
                 rb_enc_raise(utf8_encoding, cLiquidSyntaxError, "Too many filter keyword arguments");
-            }
-            if (const_keyword_args) {
-                vm_assembler_add_push_const(code, keyword_args);
+
+            if (const_keyword_args != Qnil) {
+                rb_hash_freeze(const_keyword_args);
```

Comment:
I think we are going to have to always freeze or always dup the hash.  Otherwise, it is too easy to test a filter using dynamic keyword arguments and then get errors in production when a frozen hash is passed in for constant keyword arguments.

Reply:
I've opened https://github.com/Shopify/liquid-c/pull/88 to remove the constant keyword arguments optimization on master.  We can always add it back later.